### PR TITLE
Manage multi shop images association

### LINKED ISSUE #30469: Interface
**US:**

As a merchant, I want to display specific image depending of my shops in multistore.

**Specs**
Capture d’écran 2022-10-13 à 02 32 31
The button manage image is only displayed in multistore. If pressed, it opens a modal.

Capture d’écran 2022-10-13 à 02 32 02

The modal list in a table, in row all the image in the order set by the merchant and column all the shops associated to the product.
When an image is unchecked them the image is not displayed in the FO and is disassociated to the store. 

If an image is unchecked from all the shop, then a warning is displayed saying the image will be removed. If saved, then the image is deleted from all the store.

The cover can't be unchecked.

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add one component that allows managing images associations in multishop mode
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fix https://github.com/PrestaShop/PrestaShop/issues/30469
| Related PRs       | ~
| How to test?      | Create a product in multi shop, add/upload different images in different shop contexts (switch context between the uploads) When you click on "Manage images" you should see that images are only associated to one shop and different ones You can edit the association of images on the different shops (except cover which is protected) When all associations are removed the image is removed (the dropzone content is also reset)<br>Note: The automatic cover association/re-association is not handled yet, it is part of another PR #30584 it is not the scope of this PR and shouldn't be tested
| Possible impacts? | ~

Capture d’écran 2022-12-21 à 03 24 59

### CODE REVIEW

File: admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+  const productImages = ref<ProductImage[]>([]);
+  const productShops = ref<ProductShop[]>([]);
+
+  interface ImagesShopAssociationProps {
+    productId: number;
+  }
+  const props = defineProps<ImagesShopAssociationProps>();
+
+  function closeModal(): void {
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue]: Prefer arrow functions over named functions as it reduce side effects and make it more composable

File: admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+    submittingAssociations.value = true;
+
+    // Formatted data for each image contains the list of associated shops IDs
+    const formattedAssociations = productImages.value.map((productImage: ProductImage) => ({
+      imageId: productImage.imageId,
+      shops: productImage.associations
+        .filter((association: ProductShopImage) => association.isAssociated)
+        .map((association: ProductShopImage) => association.shopId),
+    }));
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue]: Can be reduced to only one loop using `Array.reduce` increasing performances and making it memoizable way more easier

**Review by NeOMakinG — COMMENTED** [review]
Some small comments

File: admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+  const productImages = ref<ProductImage[]>([]);
+  const productShops = ref<ProductShop[]>([]);
+
+  interface ImagesShopAssociationProps {
+    productId: number;
+  }
+  const props = defineProps<ImagesShopAssociationProps>();
+
+  const closeModal = (): void => {
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue:46]: Nitpick but these return types are usually inferred and it's enough 
```suggestion
  const closeModal = () => {
```

**Review by NeOMakinG — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ImageController.php [hunk 4]
```diff
@@ -249,6 +323,10 @@ private function getErrorMessages(Exception $e): array
                 'Cannot delete file',
                 'Admin.Notifications.Error'
             ),
+            CannotRemoveCoverException::class => $this->trans(
+                'Cannot remove a cover image',
```
Comment (l-delin) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ImageController.php]:
```suggestion
                'Cannot remove cover image',
```

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ImageController.php]: Ok thanks @l-delin Fixed!

**Review by NeOMakinG — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ImageController.php [hunk 5]
```diff
@@ -198,6 +246,32 @@ private function formatImage(ProductImage $image): array
         ];
     }
 
+    private function formatShopImages(ShopProductImagesCollection $shopImagesCollection): array
```
Comment (zuk3975) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ImageController.php:254]: add php doc and define return type with psalm annotation

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Image/ImageDropzoneType.php [hunk 6]
```diff
@@ -43,6 +47,48 @@
...
+    public function __construct(
+        TranslatorInterface $translator,
+        array $locales,
+        FeatureInterface $multistoreFeature
+    ) {
+        parent::__construct($translator, $locales);
+        $this->multistoreFeature = $multistoreFeature;
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Image/ImageDropzoneType.php:67]: don't we need "@inhertiDoc" here or the return type? (I remember in cases when code inherits the return type from  older interface which has no types, we used to add that, or I think it won't break anything if you just add a return type here

**Review by zuk3975 — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Review by lartist — APPROVED** [review]

**Comment by jolelievre** [comment]
No need for functional QA on this one as this is only a partial feature, the whole image feature will be tested in the last PR that integrates cover edition for all shops correctly

**Pull request merged**